File: python/retry_safe.py
"""Retry temporary limits and failures without looping on monthly quota exhaustion."""

from __future__ import annotations

import os
import random
import time
from typing import Any

import requests

API_URL = "https://api.compsniper.com/v1/scrape"
MAX_ATTEMPTS = 5


def required_api_key() -> str:
    key = os.environ.get("COMPSNIPER_API_KEY", "").strip()
    if not key:
        raise SystemExit("Set COMPSNIPER_API_KEY before running this example.")
    return key


def json_body(response: requests.Response) -> dict[str, Any]:
    try:
        body = response.json()
        return body if isinstance(body, dict) else {}
    except requests.exceptions.JSONDecodeError:
        return {}


def delay_seconds(response: requests.Response, body: dict[str, Any], attempt: int) -> float:
    raw = response.headers.get("Retry-After") or body.get("retry_after")
    try:
        return max(0.0, float(raw)) if raw is not None else min(2**attempt, 30)
    except (TypeError, ValueError):
        return min(2**attempt, 30)


def search_sold(keyword: str) -> dict[str, Any]:
    headers = {"Authorization": f"Bearer {required_api_key()}"}

    for attempt in range(MAX_ATTEMPTS):
        try:
            response = requests.get(
                API_URL,
                headers=headers,
                params={"keyword": keyword, "count": 240},
                timeout=75,
            )
        except (requests.Timeout, requests.ConnectionError):
            if attempt == MAX_ATTEMPTS - 1:
                raise
            time.sleep(min(2**attempt, 30) + random.uniform(0, 0.5))
            continue

        body = json_body(response)
        code = body.get("code")

        if response.ok:
            print("Requests remaining:", response.headers.get("X-Usage-Remaining", "unknown"))
            return body

        if response.status_code == 429 and code == "quota_exceeded":
            raise RuntimeError(
                "Monthly quota exhausted. "
                f"Reset: {body.get('reset_at')}. Upgrade: {body.get('upgrade_url')}"
            )

        temporary = (
            response.status_code == 429 and code == "rate_limited"
        ) or response.status_code in {500, 502, 503}
        if temporary and attempt < MAX_ATTEMPTS - 1:
            wait = delay_seconds(response, body, attempt)
            time.sleep(wait + random.uniform(0, 0.5))
            continue

        message = body.get("error") or f"CompSniper returned HTTP {response.status_code}"
        raise RuntimeError(message)

    raise RuntimeError("CompSniper request failed after five bounded attempts.")


result = search_sold(os.environ.get("COMPSNIPER_KEYWORD", "9780399145636"))
print("Listings returned:", result.get("totalItems"))

File: node/retry-safe.mjs
const API_URL = "https://api.compsniper.com/v1/scrape";
const MAX_ATTEMPTS = 5;
const apiKey = process.env.COMPSNIPER_API_KEY?.trim();
if (!apiKey) throw new Error("Set COMPSNIPER_API_KEY before running this example.");

const sleep = (milliseconds) => new Promise((resolve) => setTimeout(resolve, milliseconds));

async function responseBody(response) {
  try {
    const body = await response.json();
    return body && typeof body === "object" ? body : {};
  } catch {
    return {};
  }
}

function delaySeconds(response, body, attempt) {
  const parsed = Number(response.headers.get("retry-after") ?? body.retry_after);
  return Number.isFinite(parsed) && parsed >= 0 ? parsed : Math.min(2 ** attempt, 30);
}

async function searchSold(keyword) {
  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt += 1) {
    const url = new URL(API_URL);
    url.searchParams.set("keyword", keyword);
    url.searchParams.set("count", "240");

    let response;
    try {
      response = await fetch(url, {
        headers: { Authorization: `Bearer ${apiKey}` },
        signal: AbortSignal.timeout(75_000),
      });
    } catch (error) {
      if (attempt === MAX_ATTEMPTS - 1) throw error;
      await sleep((Math.min(2 ** attempt, 30) + Math.random() * 0.5) * 1000);
      continue;
    }

    const body = await responseBody(response);
    if (response.ok) {
      console.log("Requests remaining:", response.headers.get("x-usage-remaining") ?? "unknown");
      return body;
    }

    if (response.status === 429 && body.code === "quota_exceeded") {
      throw new Error(`Monthly quota exhausted. Reset: ${body.reset_at}. Upgrade: ${body.upgrade_url}`);
    }

    const temporary =
      (response.status === 429 && body.code === "rate_limited") ||
      [500, 502, 503].includes(response.status);
    if (temporary && attempt < MAX_ATTEMPTS - 1) {
      await sleep((delaySeconds(response, body, attempt) + Math.random() * 0.5) * 1000);
      continue;
    }

    throw new Error(body.error ?? `CompSniper returned HTTP ${response.status}`);
  }

  throw new Error("CompSniper request failed after five bounded attempts.");
}

const result = await searchSold(process.env.COMPSNIPER_KEYWORD ?? "9780399145636");
console.log("Listings returned:", result.totalItems);
